tests/test_frame.py

```diff
@@ -527,9 +527,23 @@ def test_types_quantile() -> None:
     df.quantile(np.array([0.25, 0.75]))
 
 
-def test_types_clip() -> None:
+@pytest.mark.parametrize("lower", [None, 5, pd.Series([3, 4])])
+@pytest.mark.parametrize("upper", [None, 15, pd.Series([12, 13])])
+@pytest.mark.parametrize("axis", [None, 0, "index"])
+def test_types_clip(lower, upper, axis) -> None:
+    def is_none_or_numeric(val: Any) -> bool:
+        return val is None or isinstance(val, int | float)
+
     df = pd.DataFrame(data={"col1": [20, 12], "col2": [3, 14]})
-    df.clip(lower=5, upper=15)
+    uses_array = not (is_none_or_numeric(lower) and is_none_or_numeric(upper))
+    if uses_array and axis is None:
+        with pytest.raises(ValueError):
+            df.clip(lower=lower, upper=upper, axis=axis)
```

Comment:
It's possible (but not required) that you could also handle this via typing checks by creating overloads that are the acceptable combinations of the arguments `lower`, `upper` and `axis`

Reply:
> It's possible (but not required) that you could also handle this via typing checks by creating overloads that are the acceptable combinations of the arguments `lower`, `upper` and `axis`

I'm OK to approve this as is, but let me know if you want to try to make these changes I suggested as part of this PR.

Reply:
Just thinking of this implementation. One question though: the case where `upper` and `lower` are both arrays is covered by both overloads. Is that an issue ? I'm not that familiar with overloads in general

```python
    @overload
    def clip(
        self,
        lower: AnyArrayLike = ...,
        upper: float | AnyArrayLike | None = ...,
        *,
        axis: Axis = ...,
        inplace: _bool = ...,
        **kwargs,
    ) -> DataFrame: ...
    @overload
    def clip(
        self,
        lower: float | AnyArrayLike | None = ...,
        upper: AnyArrayLike = ...,
        *,
        axis: Axis = ...,
        inplace: _bool = ...,
        **kwargs,
    ) -> DataFrame: ...
```

Reply:
> I'm not sure how to type the fact that "either one of `lower` or `upper` is ArrayLike". If you have any suggestion to propose, i can add overloads

I don't think that is required.  `df.clip()` works fine, as does `df.clip(3, 4)`.  So maybe your test isn't right?


> 
> Regarding the test i think it should stay as-is, as it tests the pandas version of the function

If you can get the overloads to work, then you'd add tests using `TYPE_CHECKING_INVALID_USAGE` to make sure the type checkers caught the invalid usage.  But I'm now not sure the overloads are needed.

Reply:
> maybe your test isn't right?
Well the test is not exhaustive: `df.clip([1, 2])` will pass but `df.clip(pd.Series([1, 2])` will fail with a value error.

What's your opinion on this ? Getting dedicated overloads using pd.Series ?

Reply:
> > maybe your test isn't right?
> 
> Well the test is not exhaustive: `df.clip([1, 2])` will pass but `df.clip(pd.Series([1, 2]))` will fail with a value error.
> 
> What's your opinion on this ? Getting dedicated overloads using pd.Series ?

> > maybe your test isn't right?
> 
> Well the test is not exhaustive: `df.clip([1, 2])` will pass but `df.clip(pd.Series([1, 2]))` will fail with a value error.
> 
> What's your opinion on this ? Getting dedicated overloads using pd.Series ?

I think this is pretty complicated, because you need to have overloads for each combination where `axis=None` and using `AnyArrayLike` corresponds to `Never`, but using `AnyArrayLike` requires `axis` to be specified and not be `None`.

I'm going to approve this PR as is - it's an improvement, and if you want to experiment with the overloads, that could be a separate PR.   Created #984 to track that.